docs/core/testing/unit-testing-best-practices.md

```diff
@@ -0,0 +1,331 @@
+- **Repeatable**. Running a unit test should be consistent with its results, that is, it always returns the same result if you do not change anything in between runs.
+- **Self-Checking**. The test should be able to automatically detect if it passed or failed without any human interaction.
+- **Timely**. A unit test should not take a disproportionally long time to write compared to the code being tested. If you find testing the code taking a large amount of time compared to writing the code, consider a design that is more testable.
+
+## Lets speak the same language
+The term *mock* is unfortunately very misused when talking about testing. The following defines the most common types of *fakes* when writing unit tests:
+
+*Fake* - A fake is a generic term which can be used to describe either a stub or a mock object. Whether it is a stub or a mock depends on the context in which it's used. So in other words, a fake can be a stub or a mock.
+
+*Mock* - A mock object is a fake object in the system that decides whether or not a unit test has passed or failed. A mock starts out as a Fake until it is asserted against.
+
+*Stub* - A stub is a controllable replacement for an existing dependency (or collaborator) in the system. By using a stub, you can test your code without dealing with the dependency directly. By default, a fake starts out as a stub.
+
+Consider the following code snippet:
+
+```csharp
+var mockOrder = new MockOrder();
+var purchase = new Purchase(mockOrder);
+
+purchase.ValidateOrders();
+
+Assert.True(purchase.CanBeShipped);
+```
+
+This would be an example of Mock being used improperly. In this case, it is a stub. We're just passing in the Order as a means to be able to instantiate `Purchase` (the system under test). The name `MockOrder` is also very misleading because again, the order is not a mock.
+
+A better approach would be
+
+```csharp
+var stubOrder = new FakeOrder();
+var purchase = new Purchase(stubOrder);
+
+purchase.ValidateOrders();
+
+Assert.True(purchase.CanBeShipped);
+```
+
+By renaming the class to `FakeOrder`, we've made the class a lot more generic, the class can be used as a mock or a stub. Whichever is better for the test case. In the above example, `FakeOrder` is used as a stub. We're not using the `FakeOrder` in any shape or form during the assert. We just passed it into the `Purchase` class to satisfy the requirements of the constructor.
+
+To use it as a Mock, we could do something like this
+
+```csharp
+var mockOrder = new FakeOrder();
+var purchase = new Purchase(mockOrder);
+
+purchase.ValidateOrders();
+
+Assert.True(mockOrder.Validated);
+```
+
+In this case, we are checking a property on the Fake (asserting against it), so in the above code snippet, the `mockOrder` is a Mock.
+
+**It's important to get this terminology correct. If you call your stubs "mocks", other developers are going to make false assumptions about your intent.**
```

Comment:
We have a [style for important notes](https://github.com/dotnet/docs/blob/master/styleguide/template.md#important)